src/platform/linux/input.cpp

```diff
@@ -1583,10 +1582,51 @@ namespace platf {
    * @param input The client-specific input context.
    * @param touch_port The current viewport for translating to screen coordinates.
    * @param touch The touch event.
+   * @param input_ctx The global input context.
    */
   void
-  touch(client_input_t *input, const touch_port_t &touch_port, const touch_input_t &touch) {
-    // Unimplemented feature - platform_caps::pen_touch
+  touch(client_input_t *input, const touch_port_t &touch_port, const touch_input_t &touch, input_t &input_ctx) {
+
+    int id;
+
+    auto touchscreen = ((input_raw_t *) input_dev.get())->touch_input.get();
+    if (!touchscreen) {
+      return;
+    }
+
+
+    switch (touch.eventType)  {
+    case LI_TOUCH_EVENT_UP:
+          id = -1;
+        break;
+    case LI_TOUCH_EVENT_CANCEL:
+          id = -1;
+        break;
+    case LI_TOUCH_EVENT_CANCEL_ALL :
+        for (int i; i < 10; i++) {
+          libevdev_uinput_write_event(touchscreen, EV_ABS, ABS_MT_SLOT, i);
+          libevdev_uinput_write_event(touchscreen, EV_ABS, ABS_MT_TRACKING_ID, -1);
+        }
+        libevdev_uinput_write_event(touchscreen, EV_SYN, SYN_REPORT, 0);
+        return;
+    default:
```

Comment:
You should explicitly check for LI_TOUCH_EVENT_DOWN and LI_TOUCH_EVENT_MOVE here and drop other events.

Reply:
Do you mean adding an extra case to handle these 2 events and ? I need to know when the touch point is no longer active because it got released or cancelled as a type B uinput touchscreen doesn't care if points suddenly appear out of thin air but it wants to know when they end by signalling id -1 to the appropriate touch slot.

Reply:
No, I didn't mean dropping everything other than down/move. I meant just dropping the events you don't explicitly handle (ie the `default:` case). Rather than having `default:` write an input event (assuming the event is a down or move), instead explicitly add cases for LI_TOUCH_EVENT_DOWN and LI_TOUCH_EVENT_MOVE, then make `default:` just return.

Reply:
Done exactly as requested